File: src/game.py
from player import Player
from team import Team
from state import State, Step
from constants import Timestep
from agent import Agent, MoveDecision, ActionDecision, ActionOrResource
from logger import Logger, FileLogger, PrintLogger, LiveServerAndLogger
from network import Server
import random
from time import time
from contextlib import nullcontext

class Game:
    def __init__(self, teams, agents, seed=None):
        self.state = State.createStart(teams)
        self.teamNames = [ team.name for team in teams ]
        self.agents = agents

        if seed is None:
            seed = int(time() * 256)
            print('Using', seed, 'as master seed')
        else:
            print('Reusing seed', seed)
        random.seed(seed)

    def serialize(self):
        return { "teamNames" : self.teamNames }

    def play(self, isLiveServer : bool = False, logToTerminal : bool = False) -> bool:
        logger = Logger(self.state, self)
        logger.addSubLogger(FileLogger)

        resources = nullcontext
        if isLiveServer:
            resources = Server()
            logger.addSubLogger(LiveServerAndLogger, resources)
        if logToTerminal:
            logger.addSubLogger(PrintLogger)

        with resources:
            with logger:
                while not self.state.isFinished():
                    if self.state.timestep == Timestep.BEGIN:
                        self.agents[self.state.iActive].onBegin( self.state )
                    (self.state, step) = self.state.advance(self.agents[self.state.iActive])
                    logger.addStep(step)

File: src/game.cpp
#include "state.hpp"
#include "logger.hpp"
#include "game.hpp"
#include "network/room.hpp"
#include <random>

Generator getRandom(){
    std::random_device rd;
    return Generator(rd());
}

Game::Game(std::array<Team, 2>&& teams, const std::array<Agent*, 2>& agents, Generator generator):
    state(State::createStart(teams, generator)), teams(std::move(teams)), agents(agents)
{
};

bool Game::play(ServerRoom* serverRoom, bool logToTerminal) {
    auto* logger = new Logger(state, teams);
    if(serverRoom)
        logger = logger->liveServer(*serverRoom);
    if(logToTerminal)
        logger = logger->logToFile(std::cout);

    unsigned short int winner;
    while((winner = state.getWinner()) == 0) {
        if(state.timestep == Timestep::BEGIN)
            agents[state.iActive]->onBegin(state);
        auto [ newState, step ] = state.advance(*agents[state.iActive], *agents[1 - state.iActive]);
        state = newState;
        logger->addStep(std::move(step));
    }
    delete logger;
    return winner;
}
